xds-interop-tests: add cross branch testing to the script

sanjaypujare opened the pull request.

sanjaypujare commented:
friendly ping @sergiitk

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -44,13 +49,14 @@ readonly VERSION_TAG="master v1.50.x"
-  local clang="${3:?Usage: run_test tag server_lang client_lang}"
-  local server_image_name="${IMAGE_REPO}/${slang}-server:${tag}"
-  local client_image_name="${IMAGE_REPO}/${clang}-client:${tag}"
+  local stag="${1:?Usage: run_test server_tag client_tag server_lang client_lang}"
+  local ctag="${2:?Usage: run_test server_tag client_tag server_lang client_lang}"
+  local slang="${3:?Usage: run_test server_tag client_tag server_lang client_lang}"
+  local clang="${4:?Usage: run_test server_tag client_tag server_lang client_lang}"
```
Instead, do the check if the number of arts is 4. `${1:?}` Trick only make sense when there's a single argument.

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -44,13 +49,14 @@ readonly VERSION_TAG="master v1.50.x"
-  local clang="${3:?Usage: run_test tag server_lang client_lang}"
-  local server_image_name="${IMAGE_REPO}/${slang}-server:${tag}"
-  local client_image_name="${IMAGE_REPO}/${clang}-client:${tag}"
+  local stag="${1:?Usage: run_test server_tag client_tag server_lang client_lang}"
```
1) Can we reorder the args to group server lang with server tag, and client lang and client tag? This will read more natural
2) Tag is ambiguous here. In other scripts it make sense, because it's SHA of the commit, and refers to the docker tag. Here it's not as clear - I was thinking initially we're talking about git tags. Let's call it branch? It'll align with "cross-branch" even more.

`Usage: run_test client_lang client_branch server_lang server_branch`

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -109,15 +120,15 @@ main() {
+  # Run cross lang tests: for given cross lang versions
+  for TAG in ${XLANG_VERSIONS}
   do
     for CLANG in ${CLIENT_LANG}
     do
     for SLANG in ${SERVER_LANG}
```
We should change it to 

`for CLIENT_LANG in ${CLIENT_LANGS}`
and
`for SERVER_LANG in ${SERVER_LANGS}`

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -128,6 +139,34 @@ main() {
     done
   done
+  # Run cross branch tests per language: master x latest and master x oldest
+  for LANG in ${CLIENT_LANG}
```
Why we iterating `CLIENT_LANG` and not `SERVER_LANG`? Just curious.

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -44,13 +49,14 @@ readonly VERSION_TAG="master v1.50.x"
+  local client_image_name="${IMAGE_REPO}/${clang}-client:${ctag}"
   # TODO(sanjaypujare): skip test if image not found (by using gcloud_gcr_list_image_tags)
-  local out_dir="${TEST_XML_OUTPUT_DIR}/${tag}/${clang}-${slang}"
+  local out_dir="${TEST_XML_OUTPUT_DIR}/${ctag}-${stag}/${clang}-${slang}"
```
We might want to reverse it now, and have `${clang}-${slang}` in front. It really depends how we want to break it down in report dashboards.

sanjaypujare replied:
As discussed reversing is not required because of the structure of our dashboard.

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -22,7 +22,12 @@ readonly TEST_DRIVER_INSTALL_SCRIPT_URL="https://raw.githubusercontent.com/${TES
 readonly CLIENT_LANG="cpp go java"
-readonly VERSION_TAG="master v1.50.x"
+readonly MASTER_BRANCH="${MASTER_BRANCH:-master}"
+readonly LATEST_BRANCH_FROM_GITHUB="$((git ls-remote -t --refs https://github.com/grpc/${GITHUB_REPOSITORY_NAME}.git | cut -f 2 | sed s#refs/tags/##) | sort -V | tail -n 1 | cut -f 1-2 -d.)".x
```
Context: talked about it with @ejona86

sanjaypujare replied:
As discussed will use the version.sh file to determine the latest version. Also we decided we will test against the branch (e.g. 1.50.x) instead of a tag (e.g. 1.50.1)

sanjaypujare replied:
done

sanjaypujare commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -44,13 +49,14 @@ readonly VERSION_TAG="master v1.50.x"
+  local clang="${4:?Usage: run_test server_tag client_tag server_lang client_lang}"
+  local server_image_name="${IMAGE_REPO}/${slang}-server:${stag}"
+  local client_image_name="${IMAGE_REPO}/${clang}-client:${ctag}"
   # TODO(sanjaypujare): skip test if image not found (by using gcloud_gcr_list_image_tags)
```
done

gnossen commented on templates/VERSION.sh.template:
```diff
@@ -0,0 +1,17 @@
+# Copyright 2017 gRPC authors.
```
This should be located at `templates/internal_ci/linux/VERSION.sh`. I'm wondering how the templated file ended up in the proper place. Was that a manual move?

sanjaypujare replied:
Yes, I created it (`VERSION.sh` file) manually just to be able to test this code from kokoro. Thanks for pointing out the problem. BTW, there might be an issue with this approach - let's discuss offline.

sanjaypujare replied:
BTW, if this is to mirror the rest of the repo, shouldn't it be `templates/tools/internal_ci/linux/VERSION.sh.template` or is `tools` assumed?

gnossen replied:
Yes, it should be under tools. I was just copying from the path of your manually moved file below.

sanjaypujare replied:
All the VERSION stuff has been removed

sanjaypujare commented:
Cross branch tests passed: https://fusion2.corp.google.com/ci/kokoro/prod:grpc%2Fcore%2Fmaster%2Flinux%2Fgrpc_xds_k8s_xbranch/activity/d6a1d905-fcea-4079-876f-4fcaad7f97e9/summary

sanjaypujare commented on tools/internal_ci/linux/grpc_xds_k8s_run_test.sh:
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
```
Code shared between `grpc_xds_k8s_xlang.sh` and `grpc_xds_k8s_xbranch.sh` . Didn't want code duplication.

sanjaypujare commented on tools/internal_ci/linux/grpc_xds_k8s_xbranch.sh:
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env bash
```
Hmmm, okay. Will do that as part of another change to address this and other changes

sergiitk replied:
Did you miss this, or were planning to do it as a separate PR?

sanjaypujare replied:
Like I said, will do that as a separate PR

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_test.sh:
```diff
@@ -0,0 +1,96 @@
+  local server_image_name_tag="${server_image_name}:${server_branch}"
+  local client_image_name_tag="${client_image_name}:${client_branch}"
+
+  local out_dir="${TEST_XML_OUTPUT_DIR}/${client_branch}-${server_branch}/${client_lang}-${server_lang}"
+  mkdir -pv "${out_dir}"
```
This becomes the job name

sanjaypujare replied:
Okay, but there is nothing actionable here, right?

sanjaypujare commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -95,6 +51,14 @@ main() {
   echo "Sourcing test driver install script from: ${TEST_DRIVER_INSTALL_SCRIPT_URL}"
   source /dev/stdin <<< "$(curl -s "${TEST_DRIVER_INSTALL_SCRIPT_URL}")"
 
+# temporary commenting to test on kokoro  
```
done

sanjaypujare commented:
@sergiitk all the actionable comments have been addressed  and this PR has been rebased. Could you take a look and once approved I can merge it then?

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_test.sh:
```diff
@@ -0,0 +1,96 @@
+  go_client=$(find_latest go client)
+  java_server=$(find_latest java server)
+  java_client=$(find_latest java client)
+  LATEST_BRANCH="$((echo $cpp_server; echo $cpp_client; echo $go_server; echo $go_client; echo $java_server; echo $java_client;) | sort -g | head -1)"
```
Syntax highlighting is getting broken here. Could you please fix it? While not an issue on its own, this often indicates an underlying problem.

sergiitk replied:
You can use `shellcheck` command for linting:

```sh
In grpc_xds_k8s_run_test.sh line 30:
  LATEST_BRANCH="$((echo $cpp_server; echo $cpp_client; echo $go_server; echo $go_client; echo $java_server; echo $java_client;) | sort -g | head -1)"
                 ^-- SC1102 (error): Shells disambiguate $(( differently or not at all. For $(command substitution), add space after $( . For $((arithmetics)), fix parsing errors.
                         ^---------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                           ^---------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                             ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                                              ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                                                               ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                                                                                  ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.
```

sergiitk replied:
Also - you don't need separate echo commands. You use a single statement for it, f.e.:

```sh
❯ echo "${USER}${SHELL}"
sergiitk/bin/zsh
```

sanjaypujare replied:
`echo "${USER}\n${SHELL}"` is not guaranteed to work. But I think `echo -e "${USER}\n${SHELL}"` would work. Some of these shell idiosyncrasies are weird and that's why I went with the most basic way that's not dependent on such quirks.

sanjaypujare replied:
done

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_test.sh:
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env bash
```
Please rename this file to `grpc_xds_k8s_run_cross_test.sh` or `grpc_xds_k8s_run_xest.sh`. Current name implies it's used by other `xds_k8s` build scripts, which it's not.

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_test.sh:
```diff
@@ -0,0 +1,96 @@
+fi
+
+if [ "${OLDEST_BRANCH}" == "" ]; then
+  OLDEST_BRANCH="v1.$(expr $(echo ${LATEST_BRANCH} | cut -f 2 -d.) - 9).x"
```
Shellcheck also complains about this:

```
In grpc_xds_k8s_run_test.sh line 34:
  OLDEST_BRANCH="v1.$(expr $(echo ${LATEST_BRANCH} | cut -f 2 -d.) - 9).x"
                      ^--^ SC2003 (style): expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].
                           ^-- SC2046 (warning): Quote this to prevent word splitting.
                                  ^--------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
  OLDEST_BRANCH="v1.$(expr $(echo "${LATEST_BRANCH}" | cut -f 2 -d.) - 9).x"
```

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xbranch.sh:
```diff
@@ -0,0 +1,110 @@
+    if run_test "${LANG}" "${MAIN_BRANCH}" "${LANG}" "${LATEST_BRANCH}"; then
+      successful_string="${successful_string} ${MAIN_BRANCH}-${LATEST_BRANCH}/${LANG}"
+    else
+      failed_tests=$((failed_tests+1))
```
Use spaces around `+`: https://google.github.io/styleguide/shellguide.html#arithmetic

sanjaypujare replied:
done

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_xlang.sh:
```diff
@@ -95,6 +51,13 @@ main() {
+    exit 0
+  fi
+
+  source ${script_dir}/grpc_xds_k8s_run_test.sh
```
Double quote to prevent globbing and word splitting.
```suggestion
  source "${script_dir}/grpc_xds_k8s_run_test.sh"
```

sergiitk reviewed (changes requested):
Please sync with the latest master and address the new comments. After that, I'll need another look.

sergiitk commented:
Just realized - another thing missing was templating the latest branch as we discussed with @gnossen.

sanjaypujare commented:
> ~Just realized - another thing missing was templating the latest branch as we discussed with @gnossen.~
> 
> Update: spoke with @gnossen IRL - we should use `gcloud` command for determining the last version instead because the version is bumped in master too early for our purposes.

And that's what the code is currently doing. So nothing actionable here I am assuming (other than the other comments)

sanjaypujare commented:
> Please sync with the latest master and address the new comments. After that, I'll need another look.

Rebased to master and addressed comments. PTAL

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_xtest.sh:
```diff
@@ -0,0 +1,97 @@
+readonly IMAGE_REPO="gcr.io/grpc-testing/xds-interop"
+
+find_latest() {
+  gcloud container images list-tags --filter=tags\~v1\.\\d\\d\.x "${IMAGE_REPO}/${1}-${2}" '--format=table[no-heading](tags)' | awk -F, '{print $2}' | sort | tail -n 1
```
```suggestion
  gcloud container images list-tags --filter='tags~v1\.\d+\.x' --flatten='tags[]' --format='value(tags)' | sort --version-sort | tail -n 1
```

1. `--filter='tags~v1\.\d+\.x'` - a better filter; single quoted to  avoid escapes
2. `--flatten='tags[]' --format='value(tags)'` - just prints the matched tag name. So awk is not needed
3. `sort --version-sort` - see comment to the line 30

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_test.sh:
```diff
@@ -0,0 +1,96 @@
+  gcloud container images list-tags --filter=tags\~v1\.\\d\\d\.x "${IMAGE_REPO}/${1}-${2}" '--format=table[no-heading](tags)' | awk -F, '{print $2}' | sort | tail -n 1
+}
+
+if [ "${LATEST_BRANCH}" == "" ]; then
```
Please address this one.

sanjaypujare replied:
Moving the `source ...` statement after the `kokoro_setup_test_driver` call which achieves the same without creating a function.

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_test.sh:
```diff
@@ -0,0 +1,96 @@
+  go_server=$(find_latest go server)
+  go_client=$(find_latest go client)
+  java_server=$(find_latest java server)
+  java_client=$(find_latest java client)
```
> And that's what the code is currently doing. So nothing actionable here I am assuming

Yep. You're right. I misunderstood how this works.

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_xtest.sh:
```diff
@@ -0,0 +1,97 @@
+}
+
+if [ "${LATEST_BRANCH}" == "" ]; then
+  cpp_server=$(find_latest cpp server)
```
Do we  need to check clients and servers? I think servers should be enough. I can't imagine a situation where the server exists, and the client doesn't (for the main languages). If this happens, I'd treat it as an error that needs to be reported and fixed.

sanjaypujare replied:
I would leave the client check there. For a new release (branch) if the client image doesn't get created (for whatever reason) then this will "fall back" to the latest for which both images exist. I hope the error of a client (or server) image not getting created will be caught in a different place or process.

sanjaypujare commented:
Addressed all comments. PTAL

sergiitk commented on tools/internal_ci/linux/grpc_xds_k8s_run_xtest.sh:
```diff
@@ -0,0 +1,97 @@
+  gcloud container images list-tags --filter='tags~v1\.\d+\.x' --flatten='tags[]' --format='value(tags)' | sort --version-sort | tail -n 1
+}
+
+if [ "${LATEST_BRANCH}" == "" ]; then
```
I highly recommend to create a function. Then it's much clearer what's happening. Side-effects should be avoided as much as possible.

> Put all functions together in the file just below constants. **Don’t hide executable code between functions.** Doing so makes the code difficult to follow and results in nasty surprises when debugging.  
> If you’ve got functions, put them all together near the top of the file. **Only includes, set statements and setting constants may be done before declaring functions.**  
> — https://google.github.io/styleguide/shellguide.html#function-location

sanjaypujare replied:
done

sergiitk reviewed (approved):
This looks good with one exception: latest branch detection code *must* be a function.

When this is fixed, please merge master and re-run the tests.
